src/common/ui/codemirror-js-mixed/javascript-mixed.js

```diff
@@ -0,0 +1,361 @@
+      return style;
+    }
+
+    class Rule {
+      constructor(props) {
+        this.curContext = props.curContext;
+        // lambda for match condition
+        this.match = props.match;
+        this.nextContext = props.nextContext;
+        // optional, lambda for additional logic to run if matched
+        this.caseMatched = props.caseMatched;
+        // optional, lambda for additional logic to run if not matched
+        this.caseNotMatched = props.caseNotMatched;
+      }
+
+      run(ctx) {
+        const state = ctx.state;
+        if (this.match(ctx)) {
+          state.maybeLocalContext = this.nextContext;
+          if (state.maybeLocalContext == null) {
+            // local mode done, reset
+            state.localMode = null;
+            state.localState = null;
+          }
+          if (this.caseMatched) { this.caseMatched(ctx); }
+          return true;
+        } // case rule transition criteria not matched
+        if (this.caseNotMatched) {
+          this.caseNotMatched(ctx);
+        } else { // default not matched logic: reset local mode matching
+          state.maybeLocalContext = null;
+        }
+        return false;
+      }
+
+      // holds input parameters and return values for a rule execution
+      static createRunContext(stream, state, jsTokStyle) {
+        const ctx = {};
+
+        let _tokStr = null;
+        Object.defineProperties(ctx, {
+          stream: { value: stream, writable: false },
+          state: { value: state, writable: false },
+          jsTokStyle: { value: jsTokStyle, writable: false },
+          tokTyp: { get() { return state.jsState.lastType; } },
+          tokStr: {
+            get() {
+              if (_tokStr === null) {
+                _tokStr = stream.current();
+              }
+              return _tokStr;
+            },
+          },
+
+          // holds the output of a rule execution - the only writable property
+          tokSyle: { writable: true },
+        });
+        // somehow putting the value in defineProperties does not work
+        ctx.tokStyle = STYLE_PASS;
+        return ctx;
+      }
+    }
+
+    function matchRule(rules, stream, state, jsTokStyle) {
+      const ctx = Rule.createRunContext(stream, state, jsTokStyle);
+      for (const r of rules) {
+        if (r.curContext === (state.maybeLocalContext || '<start>')) {
+          // dbg('  rule:', r.curContext, r.match.toString());
+          const matched = r.run(ctx);
+          // dbg('  => rule output tokStyle', ctx.tokStyle);
+          if (matched) {
+            break;
+          }
+        }
+      }
+      return ctx.tokStyle;
+    }
+
+    const cssMode = CodeMirror.getMode(config, { name: 'css' });
+
+    // define the transition rules to enter local CSS mode;
+    const cssRules = [
+      // for pattern GM_addStyle(`css-string`);
+      new Rule({
+        curContext: '<start>',
+        match: c => c.tokTyp === 'variable' && c.tokStr === 'GM_addStyle',
```

Comment:
I understand the desire to make it compact but it's not a trivial parameter like e.g. `s` for a string or `c` for a character or `i` for index so `ctx` would be more appropriate especially since you're using it already in createRunContext.